### PR TITLE
Empty Columns skip type inference

### LINKED ISSUE #2657: Empty column mapped to boolean
## Description
Empty columns are set as boolean.
I found no way to change the default/fallback type to text.

## Expected behavior
I would expect using a more relaxed type, as text, since there is no information to infer the type.

## To Reproduce
Import a CSV with an empty column.

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #2657 

Given a call to the type inference function, this PR adds a check verifying that the column contains data before running the inference logic.

**Technical details**

Note that this solves #2657 in that we will infer an empty column to be of type `TEXT` when importing since we start from this type, it does _not_ change the type to `TEXT` if it starts as something else. This is because our inference logic is designed to be used starting from other types at times, and we don't want to start changing users' columns to type `TEXT` with no data. So, the logic is now:

```
(all NULL) -> no type change.
```

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `develop` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/columns/operations/infer_types.py [hunk 1]
```diff
@@ -41,26 +41,43 @@
...
-            list, return the current column type's class;
-        4. iterate through that sequence of types trying to alter the column's type to them;
-            - if the column's type is altered successfully, break iteration and return the output
-            of running infer_column_type again (trigger tail recursion);
-            - if none of the column type alterations succeed, return the current column type's
-            class.
+        0. Check for any data in the column.
+           - If the column is empty, return the column's current type
+             class.
+        0. reflect the column's type class.
```
Comment (dmos62) [inline db/columns/operations/infer_types.py:61]: 0 is a typo here I guess. Also, would like to see 1-indexed lists.

File: db/columns/operations/infer_types.py [hunk 2]
```diff
@@ -41,26 +41,43 @@
...
+           - If there's no such key on the TYPE_INFERENCE_DAG dict, or if
+             its value is an empty list, return the current column's type
+             class.
+        4. Iterate through that sequence of types trying to alter the
+           column's type to them.
+           - If the column's type is altered successfully, break
+             iteration and return the output of running infer_column_type
+             again (trigger tail recursion).
+           - If none of the column type alterations succeed, return the
+             current column's type class.
```
Comment (dmos62) [inline db/columns/operations/infer_types.py:75]: I see you've rewrapped comments from 100 to 80 characters, and changed list's capitalization and punctuation style. I don't have a problem with this style (I don't think 80 chars for docstrings is enoug though), but I think we should postpone any such formatting until we have a style/code standard. A notable exception is if a given style is significantly inconsistent, but in this case this was consistent with all of my other comments. Do you agree about postponing unurgent comment style changes?

A basic problem here is that it's bloated the diff and I'm not sure if I'm seeing all the meaningful changes I'm meant to see.

**Comment by rajatvijay** [comment]
@mathemancer you have referenced PR 2672 in the description, which seems a typo. 

> Note that this solves https://github.com/centerofci/mathesar/pull/2672 in that we will infer

File: db/columns/operations/select.py [hunk 3]
```diff
@@ -151,6 +151,17 @@ def get_column_default_dict(table_oid, attnum, engine, metadata, connection_to_u
     return {"value": default_value, "is_dynamic": is_dynamic}
 
 
+def determine_whether_column_contains_data(
+        table_oid, column_name, engine, metadata, connection_to_use=None
+):
```
Comment (dmos62) [inline db/columns/operations/select.py:156]: Unclear if this is a predicate or not. I can see that the returned variable `contains_data` is a `.scalar()`, but I'd still struggle to parse whether the return is always meant to be a boolean. Maybe rename into `does_column_contain_data` or add a docstring describing return intent?

**Review by dmos62 — CHANGES_REQUESTED** [review]
Requested changes in comments above.

File: db/columns/operations/infer_types.py (hunk 2 above)
Comment (dmos62) [reply db/columns/operations/infer_types.py:75]: Not requesting changes regarding style, btw.

**Comment by dmos62** [comment]
I was confused about there being unrelated changes here. Then I realised that this is based on https://github.com/centerofci/mathesar/pull/2671. Set merge target to it.

File: db/columns/operations/infer_types.py (hunk 2 above)
Comment (mathemancer) [reply db/columns/operations/infer_types.py:75]: To be honest, I did some punctuation changes for my own readability, and my editor did the rewrapping as per PEP8 on lines I'd changed.

Comment (mathemancer) [reply db/columns/operations/infer_types.py:75]: > A basic problem here is that it's bloated the diff and I'm not sure if I'm seeing all the meaningful changes I'm meant to see.

This is a valid complaint.

File: db/columns/operations/infer_types.py [hunk 4]
```diff
@@ -41,26 +41,43 @@
...
+        column_name,
+        engine,
+        depth=0,
+        type_inference_dag=None,
+        metadata=None,
+        columns_might_have_defaults=True,
+):
     """
-    Attempts to cast the column to the best type for it, given the mappings defined in TYPE_INFERENCE_DAG
-    and _get_type_classes_mapped_to_dag_nodes. Returns the resulting column type's class.
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply db/columns/operations/infer_types.py:47]: From [PEP 257](https://peps.python.org/pep-0257/#multi-line-docstrings):

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line. The summary line may be on the same line as the opening quotes or on the next line. The entire docstring is indented the same as the quotes at its first line (see example below).

Note _why_ the docstring summary should be a short one-liner. The docstring as it was previously breaks assumptions made by many tools that display it (such as my editor).

**Comment by mathemancer** [comment]
> @mathemancer you have referenced PR 2672 in the description, which seems a typo.
> 
> > Note that this solves #2672 in that we will infer

Good eye, thanks.

File: db/columns/operations/select.py (hunk 3 above)
Comment (mathemancer) [reply db/columns/operations/select.py:156]: I'll do the docstring. I prefer function names that make it quite obvious that they're a function in `__dict__` and so on.

File: db/columns/operations/infer_types.py (hunk 2 above)
Comment (dmos62) [reply db/columns/operations/infer_types.py:75]: If your editor didn't automatically reformat each comment or docstring you edit, that would improve the diffs.

File: db/columns/operations/infer_types.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmos62) [reply db/columns/operations/infer_types.py:45]: I don't mind having a single-line summary paragraph at the top. Might be useful to me too. This doesn't address the concern about document the return, though. I'd expect you to add a single-line paragraph at the top of the docstring in cases like this, not remove information to fit into a single line. I don't mind you removing references to other symbols there though.

**Review by mathemancer — COMMENTED** [review]
Commenting because GH makes me review my on PR for comments from the files view to appear. :(

File: db/columns/operations/infer_types.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mathemancer) [reply db/columns/operations/infer_types.py:45]: You're not wrong there. Sorry for my tone.

I think we're at an impasse w.r.t. the point and proper use of docstrings. I've been wanting to have some kind of style guide for this anyway as we try to rework some parts of the codebase. Perhaps we should bicker in other channels, then document the results somewhere when we come to a common understanding.

**Review by dmos62 — APPROVED** [review]

File: db/columns/operations/infer_types.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmos62) [reply db/columns/operations/infer_types.py:45]: No worries. I could have done a better job of drilling down to the essentials of the discussion too. As I said on Matrix, I think a lot of issues intertwined in this PR, and the difficulty to untangle that caused the friction.

**Pull request merged**